syntax = "proto3";

import "space.proto";
import "locator.proto";
import "recording.proto";

option csharp_namespace = "Gymize.Protobuf";

message ActionProto {
	string agent = 1;
	InstanceProto action = 2;
}

message ObservationProto {
	LocatorProto locator = 1;
	InstanceProto observation = 2;
}

message RewardProto {
	string agent = 1;
	double reward = 2;
}

message InfoProto {
	string agent = 1; // "" for env info
	repeated InstanceProto infos = 2;
}

message GymizeProto {
	repeated string reset_agents = 1; // "" for the environment, others for individual agents
	repeated string request_agents = 2; // ask for observations, rewards, infos or actions belong to agent name, "" will clear env info
	repeated string response_agents = 3; // agents that response to the request_agents
	repeated ActionProto actions = 4;
	repeated ObservationProto observations = 5;
	repeated RewardProto rewards = 6;
	repeated string terminated_agents = 7; // "" for the environment, others for individual agents
	repeated string truncated_agents = 8; // "" for the environment, others for individual agents
	repeated InfoProto infos = 9; // can go with observations but do not need to go with actions
	RecordingProto recording = 10; // render images and audio
}